canonical-web-and-design/deployment-configs
# Jinja2 template system

# Description

This PR will change completely the structure of our deployments configs, and we will no longer save Kubernetes objects directly in this repo.

Instead, we will have small files with variables values per each project, and we will generate the Kubernetes output when needed.

## List of changes
- Remove all Kubernetes manifests.
- Jinja templates under `templates` folder.
- All projects config values under  `sites` folder.
- konf.py script to render the templates per project
- qa-templates-checker.sh to make sure we are not breaking the current Kubernetes objects
- Updated qa-deploy to test projects locally
- Updated README.md

## New requirements

We have to be sure that the following system dependencies are installed:
- python3-jinja2
- python3-yaml

## How it works

Take a look to the updated [README.md](README.md) file to understand how konf.py works.

# QA
Download this [gist](https://gist.github.com/jkfran/fa5742f73e4b590b7ca65bb308fc939c) and place it on the root of this branch. `qa-templates-checker.sh` is a script that I created in order to verify that this template system will not produce any unwanted changes to our current configs. Feel free to take a look into it before using it.

## How the QA script works

Basically we are using `kubectl diff` to compare against all our configuration in the git branch "master". We first load all the Kubernetes objects in our system and then we do `kubectl diff` for each project.

## QA Instructions

- Make sure you have the latest version of the master branch (It's important because we do the diff using this branch)
- Do a git checkout to this branch
- run `./qa-templates-checker.sh`
- It will output all projects found and if there are changes detected or not
- Look the output and the diff.log file

## Interpret the output

We can see if there are changes by the environment (production or staging) or missing projects.
If there are changes a new file called `diff.log` has been generated where you can see these changes.

The `generation` property in the `diff.log` file can be ignored, it's going to appear always that we made a change in any Kubernetes object:
```
-  generation: 1
+  generation: 2
```

It doesn't make sense to keep this QA script in the repository after the merge but it has been necessary for the development and also to validate this PR

# Next steps

When the approval of this PR occurs, it will be necessary to start working on the necessary changes in Jenkins
